src/patchset.py

```diff
@@ -0,0 +1,322 @@
+
+        patchset_hash = self._apply_patches(checkout_path, patch_artifacts)
+        patchset_hash_short = patchset_hash[:PATCHSET_SHORT_HASH_LEN]
+
+        tarball_path = self._make_tarball(
+            target_dir=checkout_path,
+            tarball_name=f"{checkout_name}-{patchset_hash_short}"
+        )
+        tarball_url = self._push_tarball(tarball_path)
+
+        self._update_node(
+            patchset_node=patchset_node,
+            checkout_node=checkout_node,
+            tarball_url=tarball_url,
+            patchset_hash=patchset_hash
+        )
+
+    def _mark_failed(self, patchset_node):
+        node = patchset_node.copy()
+        node.update({
+            "state": "done",
+            "result": "fail",
+        })
+        try:
+            self._api.node.update(node)
+        except requests.exceptions.HTTPError as err:
+            err_msg = json.loads(err.response.content).get("detail", [])
+            self.log.error(err_msg)
+
+    def _mark_failed_if_no_parent(self, patchset_node):
+        if not patchset_node["parent"]:
+            self.log.error(
+                f"Patchset node {patchset_node['id']} as has no parent"
+                "checkout node , marking node as failed",
+            )
+            self._mark_failed(patchset_node)
+            return True
+
+        return False
+
+    def _mark_failed_if_parent_failed(self, patchset_node, checkout_node):
+        if (
+            checkout_node["state"] == "done" and
+            checkout_node["result"] == "fail"
+        ):
+            self.log.error(
+                f"Parent checkout node {checkout_node['id']} failed, "
+                f"marking patchset node {patchset_node['id']} as failed",
+            )
+            self._mark_failed(patchset_node)
+            return True
+
+        return False
+
+    def _run(self, sub_id):
+        self.log.info("Listening for new trigger events")
+        self.log.info("Press Ctrl-C to stop.")
+
+        while True:
+            patchset_nodes = self._api.node.find({
+                "name": "patchset",
+                "state": "running",
+            })
+
+            if patchset_nodes:
+                self.log.debug(f"Found patchset nodes: {patchset_nodes}")
+
+            for patchset_node in patchset_nodes:
+                if self._mark_failed_if_no_parent(patchset_node):
+                    continue
+
+                checkout_node = self._api.node.get(patchset_node["parent"])
+
+                if self._mark_failed_if_parent_failed(
+                    patchset_node,
+                    checkout_node
+                ):
+                    continue
+
+                if checkout_node["state"] == "running":
+                    self.log.info(
+                        f"Patchset node {patchset_node['id']} is waiting "
+                        f"for checkout node {checkout_node['id']} to complete",
+                    )
+                    continue
```

Comment:
Okay, now I understand why you are finding all the running nodes on every event.
According to the state machine and design, `checkout` node should only have children after its state changes to `available`.
Instead of manually creating `checkout` and `patchset` nodes, we can listen to `available checkout` node events in the patchset service and create a node automatically from the service itself.

Reply:
So, for every `patchset` node I should create a new `checkout` node and wait for it to become `available`? It will pause the queue for quite some time and seems to be quite inefficient.
However it gave me another idea. Ideally I want to apply patchset to whatever latest commit is available on a tree. Is it possible to query latest available checkout node?

Reply:
Yes, that's what I am suggesting. You can listen to available checkout nodes and create patchset nodes from the service directly.
See, for instance, you can use the below filter [here](https://github.com/yurinnick/kernelci-pipeline/blob/patchset-service/src/patchset.py#L154) to get available checkout node events:
```
def _setup(self, *args):
        return self._api_helper.subscribe_filters({
            "name": "checkout",
            "state": "available",
        })
```
Then you can get `checkout` node from the event just like https://github.com/kernelci/kernelci-pipeline/blob/main/src/tarball.py#L130 and then you can create `patchset` node from it.

Reply:
So it should be done outside of `patchset` service, right? In an another service that will prepare `checkout` node and then pass it into `patchset`?

Reply:
So I think the greatest challenge with checkout nodes so is picking the right one. I'm not sure how to determine an appropriate version to apply a patch and then correctly query a right node. If we have checkouts for `6.4`, `6.8` etc how to pick the right version. Maybe we should pick the latest stable and `next` branch and create two patchset nodes based on them.

Reply:
> In an another service that will prepare checkout node and then pass it to patchset?

The pipeline service `trigger` is already creating checkout nodes whenever a new commit is detected for a particular kernel revision. As I have not been involved in the patchset-related discussions, I am not aware of all the requirements, but do you think you can use `checkout` nodes created by the `trigger` service?
Once the trigger service creates a `checkout` node, the `tarball` service will create and push a tarball and change the node's status to `available`. Then you can use the below filter in the `patchset` service and create a patchset node from it:

```
    def _setup(self, *args):
        return self._api_helper.subscribe_filters({
            "op": "updated",
            "name": "checkout",
            "state": "available",
        })
    def _run(self, _sub_id):
        self.log.info("Listening for new trigger events")
        self.log.info("Press Ctrl-C to stop.")

        while True:
              checkout_node = self._api_helper.receive_event_node(sub_id)
              patchset_node = { create_patchset_node_from_checkout() }
              patchset_node = self._api.node.add(patchset_node)
              self._process_patchset(checkout_node, patchset_node)
```

Reply:
As we discussed with @gctucker in kernelci-pipeline#295, `patchset` service only processed patchset nodes created by another service (patchwork<>kernelci connector). So patchset service should tream these nodes as created on-demand and process them accordingly.

I am open to suggestions on how to link existing checkout nodes and freshly created patchset node. I think it's a topic for a larger discussion on how we want to test kernel patches in general.

Reply:
@JenySadadia @yurinnick I agree it will be an important next step to find the best way to connect `patchset` nodes with `checkout` ones created by the `trigger` service. Let keep the scope of this PR to just processing `patchset` nodes created by on-demand - with an assumption for an already existing parent checkout node (like in the verification steps posted in this PR).

I created #490 to track this task. @JenySadadia Please let me know if there something else that should be done before starting integration of this service.

Reply:
I think we should merge this PR now and improvements can be discussed as follow-ups.